File: EndGate/EndGate.Core/GameRunner.cs
using System;
using System.Collections.Concurrent;
using EndGate.Core.Utilities;

namespace EndGate.Core
{
    internal class GameRunner
    {
        private static Lazy<GameRunner> _instance = new Lazy<GameRunner>(() => new GameRunner());
        private ConcurrentDictionary<long, LooperCallback> _callbacks;
        private Looper _gameLoop;

        public GameRunner()
        {
            _callbacks = new ConcurrentDictionary<long, LooperCallback>();
        }

        public static GameRunner Instance
        {
            get
            {
                return _instance.Value;
            }
        }

        public Action<int> Register(Game game)
        {
            var updateCallback = CreateAndCacheCallback(game);

            // Try to start the loop prior to adding our games callback.  This callback may be the first, hence the "Try"
            TryLoopStart();

            // Add our callback to the game loop (which is now running), it will now be called on an interval dictated by updateCallback
            _gameLoop.AddCallback(updateCallback);

            // Updating the "updateRate" is an essential element to the game configuration.
            // If a game is running slowly we need to be able to slow down the update rate.
            return CreateUpdateRateSetter(updateCallback);
        }

        public void Unregister(Game game)
        {
            LooperCallback updateCallback;
            _callbacks.TryRemove(game.ID, out updateCallback);
            _gameLoop.RemoveCallback(updateCallback);

            TryLoopStop();
        }

        private void TryLoopStart()
        {
            if (_callbacks.Count == 1)
            {
                _gameLoop = new Looper();
                _gameLoop.Start();
            }
        }

        private void TryLoopStop()
        {
            if (_callbacks.Count == 0 && _gameLoop != null)
            {
                _gameLoop.Dispose();
                _gameLoop = null;
            }
        }

        private LooperCallback CreateAndCacheCallback(Game game)
        {
            var updateCallback = new LooperCallback
            {
                Callback = game.PrepareUpdate
            };

            // Add the callback to the callback cache
            _callbacks.TryAdd(game.ID, updateCallback);

            return updateCallback;
        }

        private Action<int> CreateUpdateRateSetter(LooperCallback callback)
        {
            return (updateRate) =>
            {
                callback.Fps = updateRate;
            };
        }
    }
}

File: EndGate/EndGate.Core.Client.JS/GameRunner.js
var EndGate;
(function (EndGate) {
    (function (Core) {
        var GameRunner = (function () {
            function GameRunner() {
                this._type = "GameRunner";
                this._updateCallbacks = {
                };
                this._drawCallbacks = {
                };
                this._updateLoop = null;
                this._drawLoop = null;
                this._callbackCount = 0;
            }
            GameRunner.prototype.Register = function (game) {
                var updateCallback = this.CreateAndCacheUpdateCallback(game);
                var drawCallback = this.CreateAndCacheDrawCallback(game);
                this._callbackCount++;
                this.TryLoopStart();
                this._updateLoop.AddCallback(updateCallback);
                this._drawLoop.AddCallback(drawCallback);
                return this.CreateUpdateRateSetter(updateCallback);
            };
            GameRunner.prototype.Unregister = function (game) {
                var updateCallback, drawCallback;
                if(this._updateCallbacks[game.ID]) {
                    updateCallback = this._updateCallbacks[game.ID];
                    drawCallback = this._drawCallbacks[game.ID];
                    this._updateLoop.RemoveCallback(updateCallback);
                    this._drawLoop.RemoveCallback(drawCallback);
                    delete this._updateCallbacks[game.ID];
                    delete this._drawCallbacks[game.ID];
                    this._callbackCount--;
                    this.TryLoopStop();
                }
            };
            GameRunner.prototype.TryLoopStart = function () {
                if(this._callbackCount === 1) {
                    this._updateLoop = new Core.Loopers.Looper();
                    this._updateLoop.Start();
                    this._drawLoop = new Core.Loopers.RepaintLooper();
                    this._drawLoop.Start();
                }
            };
            GameRunner.prototype.TryLoopStop = function () {
                if(this._callbackCount === 0 && this._updateLoop != null) {
                    this._updateLoop.Dispose();
                    this._updateLoop = null;
                    this._drawLoop.Dispose();
                    this._drawLoop = null;
                }
            };
            GameRunner.prototype.CreateAndCacheUpdateCallback = function (game) {
                var updateCallback = new Core.Loopers.TimedCallback(0, function () {
                    game.PrepareUpdate();
                });
                this._updateCallbacks[game.ID] = updateCallback;
                return updateCallback;
            };
            GameRunner.prototype.CreateAndCacheDrawCallback = function (game) {
                var drawCallback = new Core.Loopers.LooperCallback(function () {
                    game.PrepareDraw();
                });
                this._drawCallbacks[game.ID] = drawCallback;
                return drawCallback;
            };
            GameRunner.prototype.CreateUpdateRateSetter = function (callback) {
                return function (updateRate) {
                    callback.Fps = updateRate;
                };
            };
            return GameRunner;
        })();
        Core.GameRunner = GameRunner;        
    })(EndGate.Core || (EndGate.Core = {}));
    var Core = EndGate.Core;
})(EndGate || (EndGate = {}));
var GameRunnerInstance = new EndGate.Core.GameRunner();
